keps/sig-api-machinery/2161-apiserver-default-labels/README.md

```diff
@@ -0,0 +1,452 @@
+values for namespace selectors mean *any* namespace is allowed:
+
+|                         | old api client | new api client |
+|-------------------------|----------------|----------------|
+| namespaceAsNameSelector | ANY            | AND            |
+| namespaceSelector       | AND            | AND            |
+| ipBlock                 | NOT ALLOWED    | NOT ALLOWED    |
+
+Thus, adding a *new* namespace selector field would result in  old api clients
+assuming that ANY namespace is allowed when namespaceAsNameSelectors are utilized
+instead of the namespaceSelector fields, due to the fact that *nil
+namespaceSelectors are promiscuous*.
+
+Possible solutions to this divergent interpretation of policy might be:
+- closed failures which break on old clients (making networkpolicy api evolution very hard)
+- open failures with lots of warnings and docs about how old clients might
+overinterpret the allowed connectivity for a policy
+
+Both of these are impractical due to the nature of security tools, which need to
+be robust and explicit.  Although failing in a closed manner is acceptable, it is
+obviously unfortunate, as it would obsolete all network policy providers *not*
+eagerly adopting such a new semantic, which is overly restrictive and against
+the spirit of community-friendly, safe API evolution which is so important to
+a large project like Kubernetes with many external functional plugins.
+
+Ideally, a fix here would “just work” for “back-rev” implementations of NetworkPolicy
+(which are out-of-tree - generally implemented by cni plugins) and would not
+introduce new surprises for users. But in practicality this is not the case
+because, by adding a new selection data structure, the logical inclusion/exclusion
+properties of namespace and pod selectors become drastically different, as shown
+in the table below:
+
+|         input policy         | policy-reader (old)    | client (new)        |   |   |
+|------------------------------|------------------------|---------------------|---|---|
+| nsNameSelector + podSelector | ANY ns, pod restricted | ns + pod restricted |   |   |
+| nsNameSelector               | invalid (empty peer)   | ns restricted       |   |   |
+| podSelector                  | ANY namespace          | ANY namespace       |   |   |
+
+In this table "policy-reader" might be something like the "calico-controller"
+or any other cni agent that reads the API Servers network policy portfolio and
+responds to it over time.
+
+In an older client (one which predated the addition of a new, "nsNameSelector"
+field, we find that there is a scenario (`ANY ns, pod restricted`), wherein an
+old interpretation of the API assumes that the ONLY restricting field is a pod
+label.  Thus, although the user intended to restrict all traffic to a
+namespace/pod combination, the policy implementer does something **much** more
+promiscuous.
+
+### Improve NetworkPolicy API and deprecate old fields
```

Comment:
This seems worth exploring, with a discriminated union, you could extend this selection criteria in the future.

Reply:
Agree.  For this specific case we thought the more general solution described here would be best since it covers other users of NS selectors.